### PR TITLE
Add an option to reduce random access by caching fields of visited nodes when doing BFS over StarTree index

### PR DESCRIPTION
**Pull request by klsince**
Add a query option `reduceRandomAccess`. For now it's mainly to reduce random access when traversing StarTree index, by caching node's fields to avoid probing the underlying index buffer back and forth when moving from many parent nodes to many children nodes, thus making reads over the index buffer sequential during BFS. By default, it's off so things work as before.

### CODE REVIEW

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/startree/CachedStarTreeNode.java [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+  }
+
+  @Override
+  public boolean isLeaf() {
+    return _delegate.isLeaf();
+  }
+
+  @Override
+  public StarTreeNode getChildForDimensionValue(int dimensionValue) {
+    return new CachedStarTreeNode(_delegate.getChildForDimensionValue(dimensionValue));
```
Comment (Jackie-Jiang) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/startree/CachedStarTreeNode.java:91]: Need to check if `_delegate.getChildForDimensionValue(dimensionValue)` is `null`

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/startree/OffHeapStarTreeNode.java [hunk 2]
```diff
@@ -39,11 +39,13 @@ public class OffHeapStarTreeNode implements StarTreeNode {
   private final PinotDataBuffer _dataBuffer;
   private final int _nodeId;
   private final int _firstChildId;
+  private final int _lastChildId;
```
Comment (Jackie-Jiang) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/startree/OffHeapStarTreeNode.java:42]: Is this change required? This will add extra memory footprint to the star-tree node

File: pinot-spi/src/main/java/org/apache/pinot/spi/utils/CommonConstants.java [hunk 3]
```diff
@@ -280,6 +280,7 @@ public static class QueryOptionKey {
         public static final String SKIP_UPSERT = "skipUpsert";
         public static final String USE_STAR_TREE = "useStarTree";
         public static final String SCAN_STAR_TREE_NODES = "scanStarTreeNodes";
+        public static final String SCAN_STAR_TREE_NODES_ON_HEAP = "scanStarTreeNodesOnHeap";
```
Comment (Jackie-Jiang) [inline pinot-spi/src/main/java/org/apache/pinot/spi/utils/CommonConstants.java:283]: Should we come up with one single query option for the data locality based optimizations?

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/startree/OffHeapStarTreeNode.java (hunk 2 above)
Comment (klsince) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/startree/OffHeapStarTreeNode.java:42]: this can be avoided, by creating a new iterator in the CachedStarTreeNode instead of delegating to OffHeapStarTreeNode's iterator, when it reads lastChildId but that's already too late.

File: pinot-spi/src/main/java/org/apache/pinot/spi/utils/CommonConstants.java (hunk 3 above)
Comment (klsince) [reply pinot-spi/src/main/java/org/apache/pinot/spi/utils/CommonConstants.java:283]: maybe just the `scanStarTreeNodes` option to indicate the reads are moving forward (which indicates the use of cache to avoid looking back)

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/startree/OffHeapStarTreeNode.java (hunk 2 above)
Comment (klsince) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/startree/OffHeapStarTreeNode.java:42]: Looks like _lastChildId is needed to keep the abstraction clean to delegate calls from CachedStarTreeNode to OffHeapStarTreeNode. This extra int field doesn't add more overhead considering the object memory layout.

File: pinot-spi/src/main/java/org/apache/pinot/spi/utils/CommonConstants.java (hunk 3 above)
Comment (klsince) [reply pinot-spi/src/main/java/org/apache/pinot/spi/utils/CommonConstants.java]: called it reduceRandomAccess to be more descriptive and generic.

**Comment by richardstartin** [comment]
Out of interest how do you know this is more disk-friendly (presumably better for block prefetching and fewer page faults)? Have you looked at major faults with sar -B or something like that?

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/startree/OffHeapStarTreeNode.java (hunk 2 above)
Comment (richardstartin) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/startree/OffHeapStarTreeNode.java:42]: I’m not sure there is any alignment padding for it to fit into, I measured that here: https://github.com/apache/pinot/pull/8570#issue-1209642509

**Comment by richardstartin** [comment]
If #8570 had negative impacts for certain disk setups, I’d suggest just reverting the startreeenode changes there, rather than making good performance opt in in what could be or (end up being) a common scenario.

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/startree/OffHeapStarTreeNode.java (hunk 2 above)
Comment (klsince) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/startree/OffHeapStarTreeNode.java:42]: I counted obj ref as 8, but think I should count it as 4 bytes as UseCompressedOops is used in most cases

Comment (richardstartin) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/startree/OffHeapStarTreeNode.java:42]: For future reference, in case you haven't used it before, you can actually measure it on Hotspot using [jol](https://mvnrepository.com/artifact/org.openjdk.jol/jol-core/0.16). You just write something like:

```java
System.out.println(ClassLayout.parseClass(StarTreeNode.class).toPrintable());
```

But this takes into account you JVM settings, so if you really want to go above 32GiB heap you'd have to disable compressed oops.

**Comment by klsince** [comment]
> Out of interest how do you know this is more disk-friendly ...

This is a great question! This option is kinda experimental. We're experimenting a ReadAhead mechanism to access segments kept on remote storage via a custom SegmentDirectory object; and found that sequential access pattern would make that a lot more effective. Currently, I don't have stats about how much this option would benefit disk based setup.
